great_expectations/core/evaluation_parameters.py

```diff
@@ -208,34 +211,37 @@ def build_evaluation_parameters(
-            if not interactive_evaluation:
-                continue
-
-            # First, check to see whether an argument was supplied at runtime
-            # If it was, use that one, but remove it from the stored config
-            if "$PARAMETER." + value["$PARAMETER"] in value:
-                evaluation_args[key] = evaluation_args[key][
-                    "$PARAMETER." + value["$PARAMETER"]
-                ]
-                del expectation_args[key]["$PARAMETER." + value["$PARAMETER"]]
-
-            # If not, try to parse the evaluation parameter and substitute, which will raise
-            # an exception if we do not have a value
-            else:
-                raw_value = value["$PARAMETER"]
-                parameter_value = parse_evaluation_parameter(
-                    raw_value,
-                    evaluation_parameters=evaluation_parameters,
-                    data_context=data_context,
-                )
-                evaluation_args[key] = parameter_value
-                # Once we've substituted, we also track that we did so
-                substituted_parameters[key] = parameter_value
+    # A closure enables us to perform recursive tree traversal while sharing state with the
+    # primary function, reducing the number of parameters we need to pass to the helper function
+    def _build_evaluation_parameters(root: Any) -> None:
+        if isinstance(root, (tuple, list)):
+            for child in root:
+                _build_evaluation_parameters(child)
+        elif isinstance(root, dict):
+            for key, value in root.items():
+                if isinstance(value, dict) and "$PARAMETER" in value:
+                    # First, check to see whether an argument was supplied at runtime
+                    # If it was, use that one, but remove it from the stored config
+                    if "$PARAMETER." + value["$PARAMETER"] in value:
+                        root[key] = root[key]["$PARAMETER." + value["$PARAMETER"]]
+                        del expectation_args[key]["$PARAMETER." + value["$PARAMETER"]]
+
+                    # If not, try to parse the evaluation parameter and substitute, which will raise
+                    # an exception if we do not have a value
+                    else:
+                        raw_value = value["$PARAMETER"]
+                        parameter_value = parse_evaluation_parameter(
+                            raw_value,
+                            evaluation_parameters=evaluation_parameters,
+                            data_context=data_context,
+                        )
+                        root[key] = parameter_value
+                        # Once we've substituted, we also track that we did so
+                        substituted_parameters[key] = parameter_value
+                _build_evaluation_parameters(value)
+
+    if interactive_evaluation:
+        _build_evaluation_parameters(evaluation_args)
```

Comment:
The original logic was very much dependent on this (if it is `False`, nothing actually happens).

Moved it out so its significance is highlighted